[capnproto] Add Capnproto package

tarcila opened the pull request:
Add a new CapnProto package. Tested on Windows and Linux (Ubuntu 18.10).
The embedded patch as also been submitted upstream (https://github.com/capnproto/capnproto/pull/804)

tarcila committed:
Add Capnproto package

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=5617) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=5617) before we can accept your contribution.<br/><br/>:x: tarcila [sign now](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=5617)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vcpkg?pullRequest=5617) it.</sub>

Rastaban commented on ports/capnproto/portfile.cmake:
```diff
@@ -0,0 +1,42 @@
+    HEAD_REF master
+)
+
+vcpkg_apply_patches(
```
This can be moved into the PATCHES argument to vcpkg_from_github

tarcila replied:
👍

Rastaban commented:
This is looking good so far.  It looks like the uwp failures are expected so those can be ignored.  I also see failures arm64-windows is that expected?

tarcila committed:
Fix tools/ and add support vcpkg_test_cmake()

tarcila commented:
> This is looking good so far. It looks like the uwp failures are expected so those can be ignored. 

Yes, for uwp, the code relies on some LoadLibrary calls, so this is a no go.

> I also see failures arm64-windows is that expected?

Not sure about this one. I can't seem to have access to build logs. Anything you can share?
The original website does make mention of C++11 compatibility, a quick search did not bring more details.

tarcila commented:
I have pushed an updated version that address `PATCHES` comment, plus some cmake related cleanups.

Once https://github.com/Microsoft/vcpkg/pull/5630 is merged, this should make the vcpkg_test_cmake() statement work correctly on currently broken Linux and OSX.

Rastaban commented:
With the latest commit I see failures on everything except x86-windows.

```
Starting package 2/2: capnproto:x64-windows
Building package capnproto[core]:x64-windows... 
Could not locate cached archive: C:\vsts\_work\1\s\archives\b3\b3fd7d6ae526f020b8ca476d47d0431257398d86.zip
Building DLLs not supported. Building static instead.
-- Downloading https://github.com/capnproto/capnproto/archive/v0.7.0.tar.gz...
-- Extracting source C:/vsts/_work/1/s/downloads/capnproto-capnproto-v0.7.0.tar.gz
-- Applying patch C:/vsts/_work/1/s/ports/capnproto/0001-fix-capnpc-extension-handling-on-Windows.patch
-- Using source at C:/vsts/_work/1/s/buildtrees/capnproto/src/v0.7.0-b4a52715c4
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Performing CMake integration test
CMake Error at scripts/cmake/vcpkg_test_cmake.cmake:56 (message):
  CMake integration test failed; unable to find_package(CapnProto CONFIG
  REQUIRED)
Call Stack (most recent call first):
  ports/capnproto/portfile.cmake:39 (vcpkg_test_cmake)
  scripts/ports.cmake:71 (include)


Error: Building package capnproto:x64-windows failed with: BUILD_FAILED
Elapsed time for package capnproto:x64-windows: 3.784 min
```
Log files:
[capmproto_2484052.zip](https://github.com/Microsoft/vcpkg/files/2963785/capmproto_2484052.zip)

tarcila commented:
Thanks for the logs. Most of the failures are for now expected, waiting for a few fixes:
- #5635: CMake integration test is broken on 64 bits builds when the default CMake generator is 32 bits, hence the broken x64-windows build
- #5630: CMake targets declaration is broken for non Windows platform (binaries are moved to the tools/ folder, but <Package>Targets.cmake is not correctly updated), again breaking the CMake integration test

Checking for the logs, arm64 failures seem related to upstream code being incompatible with the target platform (__popcnt intrinsic)

tarcila commented:
@Rastaban I can remove the cmake integration test from the package. That would make the package fully functional, but only on Windows (x86 and x64). Would it be an acceptable solution?

It would later on be fixed on Linux and OSX when #5630 is integrated and cmake integration test could be added back once #5635 is fixed.

Rastaban commented:
@tarcila that sounds good to me.  Having it in and functional for some architectures is more useful than none.

tarcila committed:
Disable cmake integration test for now

Integration test is failing for now because of #5630 and #5635
This at least makes the package function on Windows x86 and x64.
